### PR TITLE
[DAG][RISCV] Use vp_reduce_* when widening illegal types for reductions

### PR DESCRIPTION
**Pull request by preames**
This allows the use a single wider operation with a restricted EVL instead of padding the vector with the neutral element.

For RISCV specifically, it's worth noting that an alternate padded lowering is available when VL is one less than a power of two, and LMUL <= m1.  We could slide the vector operand up by one, and insert the padding via a vslide1up.  We don't currently pattern match this, but we could.  This form would arguably be better iff the surrounding code wanted VL=4.  This patch will force a VL toggle in that case instead.

Basically, it comes down to a question of whether we think odd sized vectors are going to appear clustered with odd size vector operations, or mixed in with larger power of two operations.

Note there is a potential downside of using vp nodes; we loose any generic DAG combines which might have applied to the widened form.

### CODE REVIEW

File: llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp [hunk 1]
```diff
@@ -7276,6 +7305,25 @@ SDValue DAGTypeLegalizer::WidenVecOp_VECREDUCE(SDNode *N) {
   unsigned OrigElts = OrigVT.getVectorMinNumElements();
   unsigned WideElts = WideVT.getVectorMinNumElements();
 
+  // Generate a vp.reduce_op if it is custom/legal for the target.  This avoids
+  // need pad the source vector, because the inactive lanes can simply be
```
Comment (lukel97) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:7309]: the need to pad?

File: llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp [hunk 2]
```diff
@@ -7259,9 +7259,38 @@ SDValue DAGTypeLegalizer::WidenVecOp_STRICT_FSETCC(SDNode *N) {
...
+  case ISD::VP_REDUCE_SMIN:
+    return ISD::SIGN_EXTEND;
+  case ISD::VECREDUCE_UMAX:
+  case ISD::VECREDUCE_UMIN:
+  case ISD::VP_REDUCE_UMAX:
+  case ISD::VP_REDUCE_UMIN:
+    return ISD::ZERO_EXTEND;
+  }
+}
+
```
Comment (lukel97) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:7289]: Is this the same as getExtendForIntVecReduction in LegalizeIntegerTypes.cpp? If so is there a way to deduplicate it

**Review by lukel97 — APPROVED** [review]
LGTM with nits

> We could slide the vector operand up by one, and insert the padding via a vslide1up. We don't currently pattern match this, but we could. This form would arguably be better iff the surrounding code wanted VL=4. This patch will force a VL toggle in that case instead.

My mental model is that a VL toggle is always going to be cheaper than any other regular permutation/arithmetic instruction, in which case this patch would be preferable.

**Comment by preames** [comment]
I just pushed some additional changes to this review.  I had not seen Luke's LGTM when I did so, and may have made this more complicated than it should be.  Sorry!

Can I ask for a new round of review?

The major changes are:
1) Support for FP reductions.  Because of the way the VP mappings are constructed right now, this excludes VECREDUCE_FADD and VECREDUCE_FMUL (both ordered and not).
2) Remove the guard for legal mask types.  I copied that guard from load lowering originally, and the one other case I introduced was binops.  I can't come up with a way we'd need to recursively lower a mask reduce when lowering a constant mask.  :)

**Review by topperc — APPROVED** [review]
LGTM
